src/Repository.cc

```diff
@@ -594,6 +594,98 @@ bool Repository::branch_up_to_date(const std::string& branch_name)
 }
 #endif
 
+
+LibGitReference Repository::new_branch(const std::string& branch_name)
+{
+    return new_branch(branch_name, get_current_branch());
+}
+
+LibGitReference Repository::new_branch(const std::string& branch_name, const std::string& origin_branch_name)
+{
+    // checkout origin branch
+    auto ref = branch_lookup(repo_.get(), origin_branch_name, GIT_BRANCH_LOCAL);
+
+    // get latest commit
+    auto commit = get_commit(reference_name(ref.get()));
+
+    // create new branch
+    return branch_create(repo_.get(), branch_name, commit.get(), 0);
+}
+
+std::string Repository::get_current_branch()
+{
+    // get current HEAD
+    auto head = repository_head(repo_.get());
+
+    // get branch name from HEAD
+    return reference_shorthand(head.get());
+}
+
+std::vector<std::string> Repository::list_branches(int type_flag)
+{
+    git_branch_t flag;
+    if      (type_flag == 0) flag = GIT_BRANCH_ALL;
+    else if (type_flag == 1) flag = GIT_BRANCH_LOCAL;
+    else if (type_flag == 2) flag = GIT_BRANCH_REMOTE;
+    else throw Error{ cat("list_branches: unknown type_flag ", type_flag) };
+
+    std::vector<std::string> ret;
+
+    LibGitBranchIterator iter = branch_iterator(repo_.get(), flag);
+    
+    LibGitReference ref = branch_next(&flag, iter.get());
+    while(ref != nullptr)
+    {
+        ret.push_back(reference_name(ref.get()));
+        ref = branch_next(&flag, iter.get());
+    }
+    return ret;
+}
+
+void Repository::checkout(const std::string& branch_name, const std::vector<std::string>& paths)
+{
+     
+    git_checkout_options checkout_opts = GIT_CHECKOUT_OPTIONS_INIT;
+    // define the paths of files to checkout
+    checkout_opts.paths.count = paths.size();
+
+    // transform std::string input into readbale data for libgit2
+    std::vector<const char*> paths_as_cstr;
+    for(const auto& path: paths)
+        paths_as_cstr.push_back(path.c_str());
+    checkout_opts.paths.strings = const_cast<char **>(paths_as_cstr.data());
+
+    //TODO: What is the reight checkout strategy? GIT_CHECKOUT_SAFE not fitting, ..._FORCE too radical?
+    checkout_opts.checkout_strategy = GIT_CHECKOUT_FORCE;
+  
+    // find latest commit of said branch
+    auto full_branch_name = reference_name(parse_reference_from_name(repo_.get(), branch_name).get());
+    auto last_commit = get_commit(full_branch_name);
+
+    // checkout
+    auto error = git_checkout_tree(repo_.get(), (const git_object*) last_commit.get(), &checkout_opts);
```

Comment:
it is necessary. Also static_cast<> does not work for me. It gives me "invalid type conversion". Probably because it does not really know those types. C cast does not care, it converts everything.

Reply:
That is why we should not use C-style casts – you can never be sure what they do. What you need is probably a `const_cast`.

Reply:
... and in addition, they hide the declaration of `git_commit` in an implementation file so we do not have access to the `object` member. So we have no other option than using a C-style cast (or reinterpret_cast) from `git_commit*` to `git_object*`. I assume that this puts us into UB land in C++, but it will probably work nonetheless :shrug:.